### PR TITLE
Added support for `ALL` DNS lookup family

### LINKED ISSUE #4778: add support for "ALL" DNS lookup family
Envoy's [DNS lookup family setting](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-enum-config-cluster-v3-cluster-dnslookupfamily) now supports an "ALL" option. When selected, both IPv4 and IPv6 addresses are returned by the DNS resolver (if they exist), and [Happy Eyeballs](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#arch-overview-happy-eyeballs) is enabled for upstream connections, which allows Envoy to efficiently detect whether to use ipv6 or ipv4 connections. This is a nice option for a dual-stack environment where a combination of IPv4 and IPv6 may be needed, and may make a good default setting for Contour users. At a minimum, we should enable it as an option for DNS lookup family.

### PR DESCRIPTION
**Pull request by Vishal-Chdhry**
Added an `ALL` option which when selected, the DNS resolver returns both IPv4 and IPv6 addresses (if they exist),

Closes #4778 

Signed-off-by: Vishal Choudhary <contactvishaltech@gmail.com>

### CODE REVIEW

**Comment by Vishal-Chdhry** [comment]
@sunjayBhatia Are these all the changes that I need to make?

**Review by skriss — CHANGES_REQUESTED** [review]
Hey @Vishal-Chdhry, thanks for the PR! This is a good start, there are a few more places that need to be updated:

- https://github.com/projectcontour/contour/blob/main/internal/envoy/v3/cluster.go#L286
- https://github.com/projectcontour/contour/blob/main/internal/dag/httpproxy_processor.go#L465-L474
- https://github.com/projectcontour/contour/blob/main/apis/projectcontour/v1/httpproxy.go#L312-L325
- https://github.com/projectcontour/contour/blame/main/site/content/docs/main/configuration.md#L166
- https://github.com/projectcontour/contour/blame/main/site/content/docs/main/configuration.md#L370
- https://github.com/projectcontour/contour/blob/main/cmd/contour/bootstrap.go#L38
- https://github.com/projectcontour/contour/blame/main/site/content/docs/main/configuration.md#L472

Searching the codebase on "dnslookup" or "dns-lookup" should highlight any others that I've missed.

**Comment by skriss** [comment]
Tagged this one for a "minor" release note, some explanation on exactly how this works (drawing from https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-enum-config-cluster-v3-cluster-dnslookupfamily) would be helpful. In particular, mentioning "Happy Eyeballs" would be useful.

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
there is a note in the documentation of Happy Eyeballs support in Envoy:

> For clusters which use [LOGICAL_DNS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-enum-value-config-cluster-v3-cluster-discoverytype-logical-dns), this behavior is configured by setting the DNS IP address resolution policy in [config.cluster.v3.Cluster.DnsLookupFamily](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-enum-config-cluster-v3-cluster-dnslookupfamily) to the [ALL](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#envoy-v3-api-enum-value-config-cluster-v3-cluster-dnslookupfamily-all) option to return both IPv4 and IPv6 addresses.

From here: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/connection_pooling#arch-overview-happy-eyeballs

Our ExternalName and "DNS Name" clusters currently are hard-coded to STRICT_DNS cluster discovery:
- https://github.com/projectcontour/contour/blob/eaa3e5897c8cedd2a862ed3a0cf8ae78ecc1f279/internal/envoy/v3/cluster.go#L61
- https://github.com/projectcontour/contour/blob/eaa3e5897c8cedd2a862ed3a0cf8ae78ecc1f279/internal/envoy/v3/cluster.go#L191

we should check if this is the case, and if so make sure to set the cluster discovery type to LOGICAL_DNS when users select the ALL loookup family

**Comment by Vishal-Chdhry** [comment]
@sunjayBhatia did I get it right?
https://github.com/Vishal-Chdhry/contour/blob/450ec16dc09cb6c85f95d6e92e164b547681d39d/internal/envoy/v3/cluster.go#L61
https://github.com/Vishal-Chdhry/contour/blob/450ec16dc09cb6c85f95d6e92e164b547681d39d/internal/envoy/v3/cluster.go#L197

File: changelogs/unreleased/4827-Vishal-Chdhry-small.md [hunk 1]
```diff
@@ -1 +1 @@
-Don't trigger DAG rebuilds for updates/deletes of unrelated Services.
\ No newline at end of file
+Don't trigger DAG rebuilds for updates/deletes of unrelated Services.
```
Comment (sunjayBhatia) [inline changelogs/unreleased/4827-Vishal-Chdhry-small.md:1]: looks like you have different branches merged here, this change should not be included in this PR

File: internal/contour/handler.go [hunk 2]
```diff
@@ -190,6 +191,7 @@ func (e *EventHandler) onUpdate(op interface{}) bool {
 	case opUpdate:
 		if cmp.Equal(op.oldObj, op.newObj,
 			cmpopts.IgnoreFields(contour_api_v1.HTTPProxy{}, "Status"),
+			cmpopts.IgnoreFields(contour_api_v1alpha1.ExtensionService{}, "Status"),
```
Comment (sunjayBhatia) [inline internal/contour/handler.go:194]: same here, seems like a merge issue

File: internal/envoy/v3/cluster.go [hunk 3]
```diff
@@ -187,8 +193,13 @@ func DNSNameCluster(c *dag.DNSNameCluster) *envoy_cluster_v3.Cluster {
 
 	cluster.Name = envoy.DNSNameClusterName(c)
 	cluster.DnsLookupFamily = parseDNSLookupFamily(c.DNSLookupFamily)
+
+	clusterType := envoy_cluster_v3.Cluster_STRICT_DNS
+	if cluster.DnsLookupFamily == envoy_cluster_v3.Cluster_ALL {
+		clusterType = envoy_cluster_v3.Cluster_LOGICAL_DNS
+	}
 	cluster.ClusterDiscoveryType = &envoy_cluster_v3.Cluster_Type{
```
Comment (sunjayBhatia) [inline internal/envoy/v3/cluster.go:201]: nit: could use `ClusterDiscoveryType` helper here rather than constructing it ourselves

I kinda like the setup you have here vs above so we don't have to construct a `*envoy_cluster_v3.Cluster_Type` multiple times and rather just swap out the type value itself

File: internal/k8s/helpers.go [hunk 4]
```diff
@@ -44,6 +45,13 @@ func isStatusEqual(objA, objB interface{}) bool {
 				return true
 			}
 		}
+	case *contour_api_v1alpha1.ExtensionService:
```
Comment (sunjayBhatia) [inline internal/k8s/helpers.go:48]: merge issue here again

File: changelogs/unreleased/4836-Vishal-Chdry-minor.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Added support for `ALL` DNS lookup family
```
Comment (sunjayBhatia) [inline changelogs/unreleased/4836-Vishal-Chdry-minor.md:1]: would be good to add more context here around how this DNS lookup type will work, similar to what you've added to the CRD field description

**Comment by Vishal-Chdhry** [comment]
@sunjayBhatia Done!

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
so your other changelog file shouldn't be deleted from the source tree, but rather should not be included in the changes you include in this PR

I think you'll need to make sure you fetch everything from upstream `git fetch --all` and do a `git rebase origin/main` (or whatever name you've set for the remote source for this repo, if it is `upstream` then `git rebase upstream/main`) and resolve conflicts

also you'll need to make sure you have a signoff on all your commits to make sure the DCO check passes, you can do that with an interactive rebase to edit your commits `git rebase -i <reference to commit to start rebasing on`

**Comment by sunjayBhatia** [comment]
in addition seems like you need to re-run `make generate` to ensure generated YAML etc. is recreated with the appropriate content

File: changelogs/unreleased/4836-Vishal-Chdry-minor.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
```
Comment (sunjayBhatia) [inline changelogs/unreleased/4836-Vishal-Chdry-minor.md:1]: you're missing a `h` in your github username here in the file name

**Comment by Vishal-Chdhry** [comment]
I rebases this branch with `origin/main` and now it has changes from `internal/provisioner/` as well

**Comment by sunjayBhatia** [comment]
> I rebases this branch with `origin/main` and now it has changes from `internal/provisioner/` as well

Yes, I think you'll need to redo that rebase since this PR branch now has changes unrelated to this change

**Comment by Vishal-Chdhry** [comment]
@sunjayBhatia I think i should make a new fresh pr for this

**Comment by Vishal-Chdhry** [comment]
@sunjayBhatia I submitted a clean pr and i am closing this
https://github.com/projectcontour/contour/pull/4909
